### ISSUE TITLE
Kotlin inline functions are not marked as covered

### ISSUE THREAD

**Issue by cplain** [jacoco/jacoco#654]
A similar issue was raised here for Jetbrains coverage tools:
https://youtrack.jetbrains.com/issue/KT-12605
with the ultimate resolution being tracked here:
https://youtrack.jetbrains.com/issue/IDEA-185485

Although the former ticket mentions raising an issue on this project as far as I could see none has been

### Steps to reproduce
```
fun contrivedExample(condition: Boolean) {
    anInlineMethod(condition) { printLn("A thing happened") }
}

inline fun anInlineMethod(condition: Boolean, callback: () -> Unit) {
  if (condition) {
    callback.invoke()
  }
}
```

### Expected behaviour
If I write a test with contrivedExample(condition = true), anInlineMethod should be considered covered

### Actual behaviour
It isn't

JaCoCo version: 0.7.9
Operating system: macOS 10.13.2
Tool integration: Gradle

**Comment by marchof** [comment]
Some remarks:

1. When the compiler assignes correct line numbers to the inlined code source highlighting should work. `javac` does so for lambda bodies
2. If the inlined method is still present in the compiled class file we need to get some hint that the method should be excluded (See #552).
3. If the method is inlined in multiple methods and not fully executed in one of the methods this will result in partial coverage.

**Comment by yole** [comment]
Just to clarify: The main issue with line numbers and inline functions is that inline functions can be declared in different source files, so the .class file contains line numbers from different source files. This can't be expressed in a `LineNumberTable`, so Kotlin creates synthetic line numbers for lines that were inlined from other source files; those line numbers go beyond the line numbers in the original source file. The actual mapping between line numbers and original source files is stored in JSR-45 strata, which the compiler also generates.

JaCoCo needs to understand this; it's not something that can be changed on the Kotlin compiler side.

**Comment by snowe2010** [comment]
Is this any closer to being resolved?

**Comment by mformetal** [comment]
This is also an issue for me. Has there been any further steps to address this issue?

**Comment by snowe2010** [comment]
... Why were our comments marked as spam? There have been no updates on this issue for a year, it's not unreasonable to ask for an update from the maintainers. I'm not saying you have to fix anything, but marking stuff as spam seems very uncalled for.

**Comment by Godin** [comment]
@snowe2010 given that development happens publicly, GitHub does damn good job in showing all updates (see "Assignes" and "Milestones" field, linked tickets and other events, etc), so all repetative questions "any update?" that can only be followed by answer "no, all updates already shown in GitHub UI" do not add any useful information (unlike for example https://github.com/jacoco/jacoco/issues/654#issuecomment-373667831) and so were hidden.

**Comment by snowe2010** [comment]
I can clearly see the Milestones field is not filled out, therefore there is no tracking. I have also visited the other linked tickets (before commenting the first time, mind you), and they are not talking about progress on this ticket, they are talking about progress on their own tickets and just happen to be referencing this one saying that they are in fact _blocked_ on this ticket. 

I see in no way that development on this ticket is happening publicly. Either you have made a decision to work on this and _have not updated_ the Milestones/Issue/Assignees to reflect it, or have decided this is not worth doing and also not updated the Issue. Both of which are not updates. Lack of information is not information. 

In the future a better way of responding would be to state that the architecture for the problem at hand is still being worked on and to invite conversation. Anyone coming into this ticket has no clue why development has stalled or what they can do to help out. If you need help with a ticket then mark it as needing help. 

Finally, I know being an open source maintainer is hard. I maintain some small projects (that have a small user base) and I understand that you are doing this of your own good will. _*But*_, flagging comments that are requesting information doesn't help the ticket move along, it just makes people not want to help out.

**Comment by Godin** [comment]
IMO statement "I'd like to work on this ticket" sounds quite different and seems more close to initial intention than the question "is this any closer to being resolved".

Tickets #763 and #764 are not blocked on this, but opposite - they added ability to read "JSR-45 strata" which was mentioned here in https://github.com/jacoco/jacoco/issues/654#issuecomment-373667831

Absence of information literally means that there is no other information to add - there is no hidden work by us on architecture for this ticket outside of GitHub, as well as for any other open ticket. And we have not decided that this ticket is not worth to work on/look at, because in this case it will be simply closed. And yes tickets can stay open without milestone and assignee for some time because we work on this project in our spare time when we have it.

"help wanted" label can be added here, but then one can say that this is kind of unfair in respect to all other open ticket, and another can say that history shows that such label doesn't help at all.

Wanna help? Please welcome - have a look at how Kotlin compiler inlines functions, what is provided via JSR-45 and how JaCoCo performs analysis of class files to come up with an idea or implementation of how this can be solved.

Hope this addresses your question and thank you for your understanding.

**Comment by snowe2010** [comment]
WHOOPS. forgot to respond to this! haha. Here's what I had written out a month ago and not clicked send on. 

-----------------

Maybe in the future instead of asking for 'any update?' I should ask "what is required for this ticket to move forward" as usually that is exactly what I mean. I do understand that adding "help wanted" can be unfair, because maybe that deprioritizes other tickets.

**Comment by senioroman4uk** [comment]
Hi, I would like to try implementing it, if noone objects

**Comment by jhannink** [comment]
Any progress on this?

**Comment by dasAnderl** [comment]
this was actually working with kotlin < 1.4 and jacoco 0.8.5.201910111838 gradle > 6
changing kotlin to 1.4 leaves inline funcs as uncovered

**Comment by JiongBull** [comment]
IDEA code coverage supports inline function now.
![image](https://user-images.githubusercontent.com/4413502/97850434-fb7e6d80-1d2e-11eb-9fd1-20c39a2b04fc.png)

**Comment by andrejordan** [comment]
Any progress on this?

**Comment by adrianotelesc** [comment]
Any progress?

**Comment by marchof** [comment]
> Absence of information literally means that there is no other information to add - there is no hidden work by us on architecture for this ticket outside of GitHub, as well as for any other open ticket.

@adrianotelesc  Let us know what exactly is not clear about this statement. 😉

**Comment by cristan** [comment]
A workaround is to use [Kotlin/Kotlinx-Kover](https://github.com/Kotlin/kotlinx-kover) instead. It's name starts with a K, so you can be sure it's fully compatible with Kotlin 😁. The reports are compatible with JaCoCo, so it should be easy to migrate. It's very new though: currently, the latest released version is only 0.4.2.

**Comment by joost-klitsie** [comment]
I just ran into this issue as well, it is a bit annoying as it can break the Sonar quality gates. Is there anything we can do about this?

**Comment by pandelisgreen13** [comment]
any update ?

**Comment by binkley** [comment]
This is a tough card to evaluate.

One the one hand, it makes sense that JaCoCo doesn't spot this as a general JVM coverage tool with special support for Java.
On the other hand, Kotlin provides metadata bytecode class headers to help tooling like JaCoCo (also to support reflection). (I'm unclear on how this work, so I may have it wrong for a case like JaCoCo and inline functions.)
On the gripping hand, Kover is in alpha stage, so not ready for primetime in production projects.

As many before have asked, any update on this card?

**Comment by marchof** [comment]
> As many before have asked, any update on this card?

@binkley See https://github.com/jacoco/jacoco/issues/654#issuecomment-886227181

**Comment by tasomaniac** [comment]
Maybe then this should be just closed and mark as "Won't Do"

**Comment by marchof** [comment]
@jtaub A good idea how JaCoCo can be reworked to support a N:M relationship between class files and source files. JaCoCo's coverage model assumes that each class file comes from at most one source file. Inlining means that a single class file is sourced from multiple source files.

Another problem with inlining is that code is executed in classes which are not analyzed today: For example if you write unit tests the code will be inlined in the test classes itself -- which are not in scope of the coverage analysis at all.

So from the JaCoCo project side is it very unlikely that we will come up with a solution. Maybe Kotlin compiler has a compiler option to avoid inlining? Then this option can be used when building for unit testing.

**Comment by yole** [comment]
In Kotlin, inlining is a language feature that affects the semantics of the code, and not just a performance optimization. Therefore, an option to avoid inlining does not exist and cannot be added.

**Comment by marchof** [comment]
> In Kotlin, inlining is a language feature that affects the semantics of the code, and not just a performance optimization. Therefore, an option to avoid inlining does not exist and cannot be added.

In this case it is even more unlikely we can solve this on bytecode level. Maybe one day a Kotlin experts shows up here and solves it.

**Comment by Keith-Albright-Bose** [comment]
adding a vote to please address this if possible!!!  Must use inlining due to use of reified...

**Comment by yk-jemmic** [comment]
as a workaround, I'm writing a java unit test that calls these kotlin inline functions

**Comment by viva-la-v** [comment]
> as a workaround, I'm writing a java unit test that calls these kotlin inline functions

Thanks a lot, this workaround is the closest approach to elegance I found.

And since this does work, it indicates that kotlin compiler did compiled the inline functions into static functions in bytecode (at least when it's called in java). But who knows why they hide the access/visibility to it from kotlin itself. If only they did not do this.

**Comment by jtaub** [comment]
Unfortunately, calling the function from Java is not an option if you're using reified types, which is one of the main motivations for inlining a function.

**Comment by aburmeis** [comment]
@cristan, @binkley just tried Kover 0.8.3 and it has the same issue! Have a method calling Iterable.maxOf() and both (JaCoCo and Cover) tell if only have covered some of the 6 branches 😞
